### PR TITLE
Only add SSH key if it doesn't already exists

### LINKED ISSUE #14: `sshcommand acl-add dokku ` is not idempotent
If you add the same key multiple times, it will keep creating entries in .ssh/authorized_keys. I suppose that is rarely the expected behaviour so it would be nice to add a quick check to make sure the key doesn't already exist.

### PR DESCRIPTION
**Pull request by jvanbaarsen**
I'm not sure if `exit 1` is the right choice here. We might consider exiting more silently.

Fixes #14

### CODE REVIEW

**Comment by michaelshobbs** [comment]
I think `exit 1` is probably fine. Though, I've been thinking maybe we should use the key itself for primary uniqueness. 

Anybody have thoughts on this? 
cc: @progrium @josegonzalez @Flink

**Comment by michaelshobbs** [comment]
@jvanbaarsen this also needs a rebase. I added tests late last week. \o/

**Comment by jvanbaarsen** [comment]
@michaelshobbs I like the idea of using the key for uniqueness :-) Also yay for tests :D

**Comment by jvanbaarsen** [comment]
@michaelshobbs Is it also possible to run the tests locally on OSX? If not, what about we add a vagrant file so we can run them inside that container?

**Comment by jvanbaarsen** [comment]
@michaelshobbs I rebased with master. What do you propose? Merge this in and then work on uniqueness of the fingerprint? Or just change this PR to fingerprint check?

**Comment by Flink** [comment]
@michaelshobbs it might be good to check the name *and* the key, I don’t know.

File: sshcommand [hunk 1]
```diff
@@ -34,6 +34,12 @@ case "$1" in
     getent passwd "$USER" > /dev/null || false
     USERHOME=$(sh -c "echo ~$USER")
 
+    new_key=$(grep "NAME=$NAME" "$USERHOME/.ssh/authorized_keys" || true)
```
Comment (michaelshobbs) [inline sshcommand]: NOTE: This will definitely change with #17

**Comment by michaelshobbs** [comment]
@progrium any reason why the same key is currently allowed multiple times?

**Comment by progrium** [comment]
Just because it was simpler. Also IMO the user experience should be add it if it doesn't exist, do nothing if it already exists (not exit non-zero) as if it were added. If for some reason there is a use case that needs to know it already exists, that should be a separate issue and perhaps separate command.

**Comment by jvanbaarsen** [comment]
@progrium Thanks for explaining! Will update this PR later this weekend.

**Comment by michaelshobbs** [comment]
Makes sense. Quick follow-up, do you suggest 'keying' on the name, the key or both? I was thinking the actual pubkey itself. I'm not sure why you would want the same key with two different identifiers. Thoughts there?

**Comment by progrium** [comment]
Without thinking about it much, keying on name makes most sense. It's possible you could have two names/accounts/users that share a pubkey. If anything it's a non impossible edge case that would behave not as you'd expect if you tried it and it keyed on pubkey.

**Comment by michaelshobbs** [comment]
Works for me. @jvanbaarsen that work for you too?

**Comment by josegonzalez** [comment]
:+1: on @progrium's last comment. I personally would set my key on the admin user as well as my own private "jose" user, and treat them differently depending upon the ssh command.

**Comment by jvanbaarsen** [comment]
@michaelshobbs Yes! Sounds like a good solution. Tomorrow you'll receive my updated PR :)

**Comment by michaelshobbs** [comment]
Excellent. Thanks for working on this

**Comment by jvanbaarsen** [comment]
@michaelshobbs I updated the PR. Waiting for tests to pass though :)

File: sshcommand [hunk 2]
```diff
@@ -58,6 +58,12 @@ case "$1" in
 
     getent passwd "$USER" > /dev/null || false
     USERHOME=$(sh -c "echo ~$USER")
+
+    NEW_KEY=$(grep "NAME=$NAME" "$USERHOME/.ssh/authorized_keys" || true)
```
Comment (michaelshobbs) [inline sshcommand]: I don't think this works with the new quoting stuff. Would you mind adding a test to check for duplicate keys?

Comment (jvanbaarsen) [inline sshcommand]: @michaelshobbs Yeah sure :-) I'm unfamiliar with bash testing though. Can I run the test suite on OSX? Or do I have to use a Vagrant box?

Comment (michaelshobbs) [inline sshcommand]: Indeed you can, but you will need `docker-machine` to do it. I haven't documented it up yet as I just put it together last week. :)

- install https://www.docker.com/docker-toolbox
- install https://www.virtualbox.org/wiki/Downloads
- create docker machine: `docker-machine create -d virtualbox sshcommand-dev`
- source docker-machine env vars: `eval $(docker-machine env sshcommand-dev`
- run tests: `make test-in-docker`

Comment (michaelshobbs) [inline sshcommand]: Unfortunately #15 doesn't work in the docker container. I'll get a patch in shortly.

@alessio (sad panda)
```
root@box223:/app# sshcommand create sshcommand_user ls
groupadd: group 'sshcommand_user' already exists
root@box223:/app# sed -n 's#^NAME="\(.*\)"#\1#p' /etc/os-release
Debian GNU/Linux
```

**Comment by michaelshobbs** [comment]
@jvanbaarsen if that's too much, i'm happy to write the testing stuff for this. lemme know.

**Comment by jvanbaarsen** [comment]
@michaelshobbs No problem :-) I'll work on it :-) (And also add it to the readme :smile: )

**Comment by josegonzalez** [comment]
@jvanbaarsen @michaelshobbs *bump*

Once this gets done, we can make a new release (probably 0.2.0 since dokku has treated some older version as 0.1.0...).

**Comment by jvanbaarsen** [comment]
@josegonzalez Sorry, I was very busy past days. Since its new years, and I have weekend after today I make the fix! :dancers:

**Comment by jvanbaarsen** [comment]
@michaelshobbs / @josegonzalez I added the test (Backslashes in bash are weird :/ ). So once tests are green, I think this can be merged

File: sshcommand [hunk 3]
```diff
@@ -58,6 +58,12 @@ case "$1" in
 
     getent passwd "$USER" > /dev/null || false
     USERHOME=$(sh -c "echo ~$USER")
+
+    NEW_KEY=$(grep "NAME=\\\\\"$NAME"\\\\\" "$USERHOME/.ssh/authorized_keys" || true)
+    if [ ! -z "$NEW_KEY" ]; then
```
Comment (josegonzalez) [inline sshcommand]: Should we show an error here (to stderr) telling the user that the name for the key has already been taken, and that they should use another name?

**Comment by josegonzalez** [comment]
One thing I just thought about is that a single user cannot have multiple keys now.

File: sshcommand (hunk 3 above)
Comment (jvanbaarsen) [inline sshcommand]: See discussion above :-)

**Comment by jvanbaarsen** [comment]
@josegonzalez They can right? Just with a different name? Or do you mean in the context of Dokku?

File: sshcommand (hunk 3 above)
Comment (jvanbaarsen) [inline sshcommand]: Oh wait, you mean just some output and still exiting 0. I like that!

**Comment by josegonzalez** [comment]
Yeah. Basically I have 2 keys (one for my desktop and one for laptop). I would add them both to my `admin` user, and then use that `NAME` key to figure out permissioning using the `user-auth` pluginhook.

Now I have to do something a little more fancy to handle this, like assigning "name"s to roles, and authing on roles. Not sure if that's a dealbreaker, though I'd imagine lots of people just default to using the same `NAME` for many calls to `sshcommand acl-add`.

**Comment by jvanbaarsen** [comment]
@josegonzalez Hm yeah, But downside of having the same name for every key, is that removing just one key is becoming impossible. Since we do destroy on NAME, and thus handling the NAME field in a unique matter, right? In your example, you say you use the NAME to figure out permissioning, that means you have unique names right?

**Comment by josegonzalez** [comment]
That's certainly true. Not that I change the key often, but currently I go on the box and delete it manually.

I guess I'm looking at this from the viewpoint of a PaaS provider. At work, we give all our developers ssh access to servers (yes I know, no ssh access would be nice, but whatever). They can also have multiple keys - Github for Mac tends to love changing it up, and they have a home/work computer - attached to that same username.

I can delete a key if they reformat a box/it gets stolen. I can also just outright remove the user from having access if they leave the company for whatever reason.

I'm not sure what the best thing to do here.

**Comment by jvanbaarsen** [comment]
@josegonzalez Hm, ok. Now I'm starting to doubt as well :-) @michaelshobbs / @progrium What do you guys think?

**Comment by michaelshobbs** [comment]
My current use case contains unique name/key combos so I may not be the best to chime in here. I see jose's use case but given that the name is easier to change (potentially) than the key, making that the unique attribute seems to make the most sense. 

@josegonzalez in your use case then couldn't you add two keys with the names `savant-work` and `savant-home`?

Then the question remains, do we error on a duplicate name or overwrite it with additional output? I'm leaning towards `exit 1` with a message in the case of a duplicate name.

**Comment by jvanbaarsen** [comment]
@michaelshobbs I don't think exiting with non-zero is needed here. We can however output a warning message, and still exit with 0?

**Comment by josegonzalez** [comment]
@michaelshobbs we could, but then if you want to handle the concept of a user, you'd need to handle that outside of dokku. Sad panda I guess.

**Comment by michaelshobbs** [comment]
@josegonzalez do you have a suggestion then on how to proceed?

**Comment by josegonzalez** [comment]
We should leave the deletion strategy as is I guess. It's a primitive any higher-level system will need to deal with.

We should def show an error message on stderr and then exit non-zero.

**Comment by michaelshobbs** [comment]
@jvanbaarsen that work for you?

**Comment by jvanbaarsen** [comment]
:+1: Will update!

**Comment by jvanbaarsen** [comment]
@michaelshobbs Pushed update

File: sshcommand [hunk 4]
```diff
@@ -58,6 +58,13 @@ case "$1" in
...
+      echo "Duplicated SSH Key name"
```
Comment (michaelshobbs) [inline sshcommand]: Should probably say "Duplicate". @josegonzalez mentioned printing this to stderr but we don't do that pattern anywhere else. @josegonzalez what was your thinking there?

**Comment by michaelshobbs** [comment]
Looks good. Would you mind squashing this into a single commit? Please and thank you. :smile:

File: sshcommand (hunk 4 above)
Comment (josegonzalez) [inline sshcommand]: We should start using this pattern and use this pattern everywhere if possible (though that can be a separate commit). Follow the unix way :)

**Comment by jvanbaarsen** [comment]
@michaelshobbs Yes will do :)

**Comment by jvanbaarsen** [comment]
@michaelshobbs Done

**Comment by michaelshobbs** [comment]
The current PR doesn't seem to reflect the changes discussed regarding echoing to stderr and a slight adjustment to the message output. :grimacing:

**Comment by jvanbaarsen** [comment]
@michaelshobbs oh haha xD I was handling a lot of PRs, and I only read the part about squashing :dancers: My bad! Will make the other changes tomorrow

**Comment by michaelshobbs** [comment]
No worries. Thanks for doing the work!!

**Comment by jvanbaarsen** [comment]
@michaelshobbs made the changes :)

**Comment by michaelshobbs** [comment]
Thanks :+1:

**Comment by jvanbaarsen** [comment]
Woohoo :dancers:

**Comment by mixxorz** [comment]
This still exits non-zero on duplicate key. Is this what we wanted?

**Comment by michaelshobbs** [comment]
Yes. See @josegonzalez comment
